applications/FluidDynamicsApplication/custom_processes/embedded_nodes_initialization_process.h

```diff
@@ -0,0 +1,334 @@
+            for (int k = 0; k < static_cast<int>(rNodes.size()); ++k)
+            {
+                unsigned int NewNodes = 0;
+                ModelPart::ElementsContainerType::iterator itElement = rElements.begin() + k;
+                const GeometryType& rGeometry = itElement->GetGeometry();
+                const unsigned int ElemNumNodes = rGeometry.PointsNumber();
+
+                // Get the number of nodes that have switched from structure to fluid in the current element
+                for (unsigned int j=0; j<ElemNumNodes; ++j)
+                {
+                    if (rGeometry[j].Is(SELECTED))
+                        NewNodes++;
+                }
+
+                // If there is only one unique "new" node in the element, initialize it.
+                // Otherwise it remains to be initialized in a further iteration.
+                if (NewNodes == 1)
+                {
+                    NodeType::Pointer pNode;
+                    double p_avg = 0.0;
+                    array_1d<double, 3> v_avg;
+                    v_avg[0] = 0.0;
+                    v_avg[1] = 0.0;
+                    v_avg[2] = 0.0;
+
+                    for (unsigned int j=0; j<ElemNumNodes; ++j)
+                    {
+                        if (rGeometry[j].IsNot(SELECTED))
+                        {
+                            // Compute the average velocity in the non selected nodes
+                            p_avg += rGeometry[j].FastGetSolutionStepValue(PRESSURE);
+                            v_avg += rGeometry[j].FastGetSolutionStepValue(VELOCITY);
+                        }
+                        else
+                        {
+                            // Get a pointer to the unique SELECTED node
+                            NodeType::Pointer pAux = rGeometry(j);
+                            std::swap(pAux, pNode);
+                        }
+                    }
+
+                    // Compute the non-SELECTED nodes average values
+                    p_avg /= (ElemNumNodes-1);
+                    v_avg /= (ElemNumNodes-1);
+
+                    // Historical values initialization
+                    pNode->SetLock();                                               // Avoid another thread to acces the node while the value set
+                    pNode->Set(SELECTED, false);                                    // Once a node has been initialized it is marked as non SELECTED
+                    for (unsigned int step=0; step<BufferSize; ++step)              // Fill the velocity and pressure buffer
+                    {
+                        pNode->FastGetSolutionStepValue(PRESSURE, step) = p_avg;
+                        pNode->FastGetSolutionStepValue(VELOCITY, step) = v_avg;
+
+                    }
+                    pNode->UnSetLock();                                             // Free the node thread access
+                }
+            }
+        }
+
+        array_1d<double, 3> v_null;
```

Comment:
your code is correct, but you can write

     array_1d<double, 3> v_null = ZeroVector(3);